### ISSUE TITLE
Unclear how "simple bound" is defined when typedefs are in use

### ISSUE THREAD

**Issue by stereotype441** [dart-lang/sdk#34722]
https://github.com/dart-lang/sdk/blob/master/docs/language/informal/instantiate-to-bound.md defines a notion of a "simple bound", but I'm unclear on how the definition applies when typedefs are in use.

Example 1:
```
class C<T extends C Function()> {}
```
In this case I think the spec is clear: `T`'s bound is not simple, because it refers back to `C`.  Therefore the appearance of `C` without explicit type parameters in the bound of `T` is illegal.

Example 2:
```
class C<T extends F> {}
typedef C F();
```
Is this ok or not?  On IM, @leafpetersen proposed that this should be invalid, and proposed the following rules:
> - If B is a bound, then consider every type T that occurs in B.
> - If T is a parameterized class, T must have simple bounds.
> - If T is a parameterized typedef, then T must have simple bounds
> - If T is a typedef (including parameterized), then all of the sub terms of its RHS must also be simple.

Example 3:
```
class C<T extends F<C>> {}
typedef void F<T>();
```
Is this ok or not?

I would prefer if we define it to be invalid (since this makes the implementation easier).  In other words, we consider `C` to "occur in" `F<C>` even though it does not occur in the expansion of `F<C>`.  Based on our IM I believe Leaf is leaning this direction as well.

@eernstg @lrhn

**Comment by leafpetersen** [comment]
To expand on the above, I think the intention is as follows.

If a type variable `Xj` has bound `Bj`, then `Xj` has a **simple** bound if both:
  - The `Xi` do not occur in `Bj`
  - Every type `T` which occurs (syntactically) in `Bj` is **shallow**

A type `T` is **shallow** if either:
  - `T` is not a parameterized type, and is not a typedef
  - `T` is a parameterized class with type arguments provided
  - `T` is a parameterized class  `C` with type arguments elided, and `C` has **simple** bounds.
  - `T` is a typedef with no parameters or for which type arguments are provided, and every `S` which occurs syntactically in the RHS of the definition of `T` is **shallow**
  - `T` is a parameterized typedef without provided type arguments, and both:
    - `T` has **simple** bounds
    - Every `S` which occurs syntactically in the RHS of the definition of `T` (with the parameters to `T` replaced with their simple bounds) is **shallow**.

**Comment by eernstg** [comment]
I'm actually not 100% sure whether the two definitions are equivalent, but let me try to retrace the answers as I think we should get them from the feature spec about [instantiate-to-bound](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/instantiate-to-bound.md):

```dart
class C<T extends C Function()> {}
```
Agreed, this one is clear: `C` occurs raw in a bound, so `C` must have simple bounds; in particular `C Function()` will be the _B<sub>j</sub>_ in the definition of simple bounds, bullet 2, and _T_ will be `C`, so `C` must have simple bounds, cycle: "No, `C` does not have simple bounds". So it's an error.

```dart
class C<T extends F> {}
typedef C F();
```
The feature spec does not make this explicit, but no other approach makes sense than considering a type alias as the same thing as the underlying function type. In particular: `C` occurs raw in the bound of `T`, because that bound is just `C Function()`, and it doesn't matter that we spelled it "F". This case is now identical to the previous one, and we have an error.

We know that type aliases can be expanded eagerly because they do not admit recursion.

I've [noted](https://github.com/orgs/dart-lang/projects/14#card-13698205) that this treatment of type aliases should be made explicit in dartLangSpec.tex when I integrate 'instantiate-to-bound.md'.

```dart
class C<T extends F<C>> {}
typedef void F<T>();
```
Expanding `F` eagerly, this is the same thing as the following:
```dart
class C<T extends void Function()> {}
```
This is then not an error. It is a corner case because unused type parameters in a parameterized type alias are utterly useless.

(We don't abstract over type aliases, so it is never required for `F` to have the given type parameter due to some first-class usage where "a type alias taking one type parameter" is required, unlike functions that might need to have a parameter which is ignored so as to get the required function type for some usage).

I think the main thing here is that the treatment of type aliases is currently underspecified, but I also think we can decide on eagerly expanding them, in which case the answers fall out easily.

(And I still won't try to conclude anything about the equivalence with **shallow** types. ;-)

**Comment by stereotype441** [comment]
Note: in https://dart-review.googlesource.com/c/sdk/+/78867 I've tried to make the analyzer consistent with the behavior in Leaf's comment.  I believe that this is the same as what Erik wants except for the treatment of unused type parameters in parameterized type aliases (which, as Erik argues, is a corner case because unused type parameters in a parameterized type alias are utterly useless).

If we decide to go with Erik's approach, we should open a bug to remind us to fix the analyzer's behavior.

**Comment by eernstg** [comment]
[This CL](https://dart-review.googlesource.com/c/sdk/+/78982) clarifies [instantiate-to-bound.md](https://github.com/dart-lang/sdk/blob/master/docs/language/informal/instantiate-to-bound.md) to specify that "`B` _contains_ a raw type `T`" is true both when `T` is a syntactic subterm of `B` and when `B` has `F` as a subterm where `F` denotes a type alias whose body contains `T` (and that second occurrence of 'contains' ensures recursive checks on further type aliases in that body).

It is slightly different from my proposed approach above, because it does not actually expand all type aliases in the bound; this means that `F<C>` would still cause `C` to be taken into consideration also in the case where the type argument of `F` is unused, and it means that type alias type bounds are checked just like generic class type bounds, in the usual recursive traversal. It just specifies that we must _also_ check that raw types in type alias bodies have simple bounds, whenever we encounter a type alias.

Note that (importantly!), we keep simple bounds simple, and in particular we don't rely on instantiate-to-bounds in the definition of simple bounds, and the reason why this is possible is that we don't have to worry about the choice of actual type arguments when we encounter a raw usage of a parameterized type alias _F_ in a bound: We don't care what actual type arguments are used, because this makes no difference for whether or not the body of _F_ contains any specific raw type.

About the notion of shallow types:

> A type `T` is **shallow** if either:
>   - `T` is not a parameterized type, and is not a typedef
>   - `T` is a parameterized class with type arguments provided
>   - `T` is a parameterized class  `C` with type arguments elided, and `C` has **simple** bounds.
>   - `T` is a typedef with no parameters or for which type arguments are provided, and every `S` 
>     which occurs syntactically in the RHS of the definition of `T` is **shallow**
>   - `T` is a parameterized typedef without provided type arguments, and both:
>       - `T` has **simple** bounds
>       - Every `S` which occurs syntactically in the RHS of the definition of `T` (with
>         the parameters to `T` replaced with their simple bounds) is **shallow**.

If I understand this correctly then I'd want to make a couple of small adjustments in terminology.

First, I think 'is not a parameterized type, and is not a typedef' should not be applicable to `List`, which is not a parameterized type in the usual terminology of the language specification (because no actual type arguments are passed). I think that `List` is intended to be covered by bullet 3 instead. So maybe the first bullet should be '`T` is a `qualified` term that denotes a non-generic type which is not a type alias'? This includes non-generic classes and oddballs like `dynamic`, and excludes raw types. Similarly, I think 'parameterized class' is the same thing as 'generic class' in the spec, in the following bullets.

The very last bullet uses 'simple bounds' as if they were specific type expressions, but we actually only use 'simple bounds' to characterize the bounds in a list of declarations of type parameters, so 'replacing parameters by their simple bounds' doesn't really make sense. I suspect that this would require us to use instantiate-to-bound in the definition of simple bounds, and that's a level of recursion (and hence complexity) in the notion of simple bounds that I'd like to avoid.

The way I did that was with a definition that requires the 'simple bounds' check to visit the body of each type alias `F` encountered during the recursive traversal that we've used all the time (that is, when we check whether `<X extends C, Y extends List<D>>` has simple bounds, we'll check the type parameter declaration list of `C` and `D`, recursively), but I ignore the actual type arguments of `F` because _all_ admissible choices of actual type arguments will have exactly the same set of raw types in the body of `F`, and we can detect which raw types that is using syntactic criteria alone.

So, modulo a couple of terminology differences I think we're essentially saying the exact same thing, but I think the definition given in [said CL](https://dart-review.googlesource.com/c/sdk/+/78982) is simpler than the notion of shallow types. I don't see them contradicting each other, though.
